applications/welcome/models/access_control.py

```diff
@@ -0,0 +1,61 @@
+#########################################################################
+## This file defines access control methods, including verification, setting and deleting.
+## - access_view can view only.
+## - access_setting can change settings, and view variables/output, but cannot change procedure code.
+## - access_manage can do everything the main user can do.
+#########################################################################
+
+
+#########################################################################
+# Verification methods
+#########################################################################
+
+# return if the user has authority to view device
+def access_view(user_id, device_token):
+    return auth.has_permission(name="view", group_id=(device_token + "_" + user_id), user_id=user_id)
+
+
+# return if the user has authority to configure device
+def access_setting(user_id, device_token):
+    return auth.has_permission(name="setting", group_id=(device_token + "_" + user_id), user_id=user_id)
+
+
+# return if the user has authority to share the device with others
+def access_manage(user_id, share_user_email, device_token):
+    return auth.has_permission(name="manage", group_id=(device_token + "_" + user_id), user_id=user_id)
+
+
+#########################################################################
+# Setting methods
+#########################################################################
+
+# give user_id "view" access to "device_token"
+def set_access_view(user_id, device_token):
+    return auth.add_permission(name="view", group_id=(device_token + "_" + user_id), user_id=user_id)
+
```

Comment:
Correct me if I am wrong. "auth.add_permission(name="view", group_id=device_token, user_id=user_id)" may not work, because add_permission does not have the parameter "user_id". In this situation, permission is given according to their group_id, so I use the combination of device and user_id as group id. I think it would be better if we set up groups according to the relationship between device and user. Then we can directly set permission to a user for a specific device. 
I will add more comments in next commit.